### ISSUE TITLE
Android client crashes on startup

### ISSUE THREAD

**Issue by open-dynaMIX** [badaix/snapcast#365]
Today I've updated snapcast server and my clients from v0.11.0 to v0.13.0. Everything works for the newer phones (android 6.0.1 & 7.0). On the older/cheap phones with android 4.4.2 (I bought them just for using them as snapcast clients :smile: ), the app instantly crashes on startup. After downgrading to v0.12.0 everything works.

Thank you very much for your work on snapcast!

**Comment by open-dynaMIX** [comment]
Sorry for not providing more information in my initial issue. I'm not experienced with debugging android apps.

I now took the time to setup adb and this is the log from the crash:

```
E/        (  660): Could not open '/data/data/hotplug/cmd'
E/        (  660): error : 2, No such file or directory
E/HwLauncher( 1053): Launcher dialog dismiss failed : java.lang.IllegalArgumentException: no dialog with id 1 was ever shown via Activity#showDialog
E/dalvikvm(32291): Could not find class 'android.graphics.drawable.RippleDrawable', referenced from method android.support.v7.widget.AppCompatImageHelper.hasOverlappingRendering
E/Setup   (32291): Failed to copy asset file: bin/armeabi-v7a/snapclient
E/Setup   (32291): java.io.FileNotFoundException: bin/armeabi-v7a/snapclient
E/Setup   (32291):  at android.content.res.AssetManager.openAsset(Native Method)
E/Setup   (32291):  at android.content.res.AssetManager.open(AssetManager.java:329)
E/Setup   (32291):  at android.content.res.AssetManager.open(AssetManager.java:303)
E/Setup   (32291):  at de.badaix.snapcast.utils.Setup.copyAsset(Setup.java:77)
E/Setup   (32291):  at de.badaix.snapcast.utils.Setup.copyBinAsset(Setup.java:62)
E/Setup   (32291):  at de.badaix.snapcast.MainActivity$2.run(MainActivity.java:149)
E/Setup   (32291):  at java.lang.Thread.run(Thread.java:841)
E/WifiManager(  660): calculateSignalLevel rssi:-57,numLevels:5,signalLevel:4
E/        (32291): appName=de.badaix.snapcast, acAppName=/system/bin/surfaceflinger
E/        (32291): 0
E/        (32291): appName=de.badaix.snapcast, acAppName=/system/bin/surfaceflinger
E/        (32291): 0
E/AndroidRuntime(32291): FATAL EXCEPTION: main
E/AndroidRuntime(32291): Process: de.badaix.snapcast, PID: 32291
E/AndroidRuntime(32291): android.view.InflateException: Binary XML file line #24: Error inflating class ImageButton
E/AndroidRuntime(32291):    at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:713)
E/AndroidRuntime(32291):    at android.view.LayoutInflater.rInflate(LayoutInflater.java:755)
E/AndroidRuntime(32291):    at android.view.LayoutInflater.inflate(LayoutInflater.java:492)
E/AndroidRuntime(32291):    at android.view.LayoutInflater.inflate(LayoutInflater.java:397)
E/AndroidRuntime(32291):    at android.view.LayoutInflater.inflate(LayoutInflater.java:353)
E/AndroidRuntime(32291):    at de.badaix.snapcast.ClientItem.<init>(ClientItem.java:54)
E/AndroidRuntime(32291):    at de.badaix.snapcast.GroupItem.update(GroupItem.java:96)
E/AndroidRuntime(32291):    at de.badaix.snapcast.GroupItem.setGroup(GroupItem.java:146)
E/AndroidRuntime(32291):    at de.badaix.snapcast.GroupItem.<init>(GroupItem.java:85)
E/AndroidRuntime(32291):    at de.badaix.snapcast.GroupListFragment$GroupAdapter.getView(GroupListFragment.java:139)
E/AndroidRuntime(32291):    at android.widget.AbsListView.obtainView(AbsListView.java:2374)
E/AndroidRuntime(32291):    at android.widget.ListView.makeAndAddView(ListView.java:1812)
E/AndroidRuntime(32291):    at android.widget.ListView.fillDown(ListView.java:698)
E/AndroidRuntime(32291):    at android.widget.ListView.fillFromTop(ListView.java:759)
E/AndroidRuntime(32291):    at android.widget.ListView.layoutChildren(ListView.java:1631)
E/AndroidRuntime(32291):    at android.widget.AbsListView.onLayout(AbsListView.java:2185)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:515)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.onLayout(FrameLayout.java:450)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.support.design.widget.CoordinatorLayout.layoutChild(CoordinatorLayout.java:1171)
E/AndroidRuntime(32291):    at android.support.design.widget.CoordinatorLayout.onLayoutChild(CoordinatorLayout.java:856)
E/AndroidRuntime(32291):    at android.support.design.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:875)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:515)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.onLayout(FrameLayout.java:450)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1914)
E/AndroidRuntime(32291):    at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1768)
E/AndroidRuntime(32291):    at android.widget.LinearLayout.onLayout(LinearLayout.java:1677)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:515)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.onLayout(FrameLayout.java:450)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1914)
E/AndroidRuntime(32291):    at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1768)
E/AndroidRuntime(32291):    at android.widget.LinearLayout.onLayout(LinearLayout.java:1677)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:515)
E/AndroidRuntime(32291):    at android.widget.FrameLayout.onLayout(FrameLayout.java:450)
E/AndroidRuntime(32291):    at android.view.View.layout(View.java:15133)
E/AndroidRuntime(32291):    at android.view.ViewGroup.layout(ViewGroup.java:4872)
E/AndroidRuntime(32291):    at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2317)
E/AndroidRuntime(32291):    at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2023)
E/AndroidRuntime(32291):    at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1189)
E/AndroidRuntime(32291):    at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6223)
E/AndroidRuntime(32291):    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:788)
E/AndroidRuntime(32291):    at android.view.Choreographer.doCallbacks(Choreographer.java:591)
E/AndroidRuntime(32291):    at android.view.Choreographer.doFrame(Choreographer.java:560)
E/AndroidRuntime(32291):    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:774)
E/AndroidRuntime(32291):    at android.os.Handler.handleCallback(Handler.java:808)
E/AndroidRuntime(32291):    at android.os.Handler.dispatchMessage(Handler.java:103)
E/AndroidRuntime(32291):    at android.os.Looper.loop(Looper.java:193)
E/AndroidRuntime(32291):    at android.app.ActivityThread.main(ActivityThread.java:5342)
E/AndroidRuntime(32291):    at java.lang.reflect.Method.invokeNative(Native Method)
E/AndroidRuntime(32291):    at jav
E/        (  660): Could not open '/data/data/hotplug/cmd'
E/        (  660): error : 2, No such file or directory
E/ActivityManager(  660): Service ServiceRecord{432dcae8 u0 de.badaix.snapcast/.SnapclientService} in process ProcessRecord{420db9a0 32291:de.badaix.snapcast/u0a161} not same as in map: null
E/InputDispatcher(  660): channel '4212b158 de.badaix.snapcast/de.badaix.snapcast.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
E/WifiManager(  660): calculateSignalLevel rssi:-57,numLevels:5,signalLevel:4
```

This was the output when running `adb logcat \*:E`. If there is a better way to obtain the relevant information, please let me know.

**Comment by badaix** [comment]
I cannot see what went wrong from this stack trace. Unfortunately, the trace is cut at `E/AndroidRuntime(32291):    at jav`.
Also I could not reproduce the crash in the emulator with API level 16.
From the history point of view the most suspicious commit is the update of my local SDK: 8645643a165ed7a1073420fd0e3a506fbfdae704

**Comment by open-dynaMIX** [comment]
I've tried it again and the trace is always cut at exactly the same position.

[Some googling revealed, that Android 4.4.2 (KitKat) in fact uses API level 19](https://developer.android.com/about/versions/android-4.4.html). Again, I'm no android dev, so you could've had good reasons to try with API level 16.

Thanks for your time.

**Comment by badaix** [comment]
I tool API level 16 because it is the lowest supported API.
The crash might be related to AppCompat v23.2 and [VectorDrawables](https://chris.banes.me/2016/02/25/appcompat-vector/#enabling-the-flag), as also reported and discussed [here](https://github.com/pockethub/PocketHub/issues/995).
I changed the code accordingly to be compatible with API level < 21. Please try this APK:
[Snapcast_0.13.1.apk.zip](https://github.com/badaix/snapcast/files/1804377/Snapcast_0.13.1.apk.zip)

**Comment by open-dynaMIX** [comment]
Thank you very much!

The app runs now, but the phone doesn't show up as a client and when pressing "play", there is no sound. I can control other clients though.

**Comment by badaix** [comment]
Argh, I remember... I've updated to NDK level 21, because I stumbled into [this problem](https://stackoverflow.com/questions/44736135/ndk-clang-error-undefined-reference-to-localeconv).
For Android the json.hpp is already patched and now I did some more patching to get it compiled with NDK level 16. Please try again: [Snapcast_0.13.1.apk.zip](https://github.com/badaix/snapcast/files/1804601/Snapcast_0.13.1.apk.zip)

**Comment by gustavokch** [comment]
Hello, badaix.

I'm having the same issue using an Oreo rom (LineageOS 15.1), even with the latest apk you posted here. v0.12 fixes the problem for me as well.

Here is the logcat of the exception:

03-20 16:41:49.636 32365  1721 E Setup   : Failed to copy asset file: bin/arm64-v8a/snapclient
03-20 16:41:49.636 32365  1721 E Setup   : java.io.FileNotFoundException: bin/arm64-v8a/snapclient
03-20 16:41:49.636 32365  1721 E Setup   :      at android.content.res.AssetManager.openAsset(Native Method)
03-20 16:41:49.636 32365  1721 E Setup   :      at android.content.res.AssetManager.open(AssetManager.java:388)
03-20 16:41:49.636 32365  1721 E Setup   :      at android.content.res.AssetManager.open(AssetManager.java:362)
03-20 16:41:49.636 32365  1721 E Setup   :      at de.badaix.snapcast.utils.Setup.copyAsset(Setup.java:77)
03-20 16:41:49.636 32365  1721 E Setup   :      at de.badaix.snapcast.utils.Setup.copyBinAsset(Setup.java:62)
03-20 16:41:49.636 32365  1721 E Setup   :      at de.badaix.snapcast.MainActivity$2.run(MainActivity.java:149)
03-20 16:41:49.636 32365  1721 E Setup   :      at java.lang.Thread.run(Thread.java:764)
03-20 16:41:49.651 32365  1721 E Setup   : Failed to copy asset file: bin//snapclient
03-20 16:41:49.651 32365  1721 E Setup   : java.io.FileNotFoundException: bin//snapclient
03-20 16:41:49.651 32365  1721 E Setup   :      at android.content.res.AssetManager.openAsset(Native Method)
03-20 16:41:49.651 32365  1721 E Setup   :      at android.content.res.AssetManager.open(AssetManager.java:388)
03-20 16:41:49.651 32365  1721 E Setup   :      at android.content.res.AssetManager.open(AssetManager.java:362)
03-20 16:41:49.651 32365  1721 E Setup   :      at de.badaix.snapcast.utils.Setup.copyAsset(Setup.java:77)
03-20 16:41:49.651 32365  1721 E Setup   :      at de.badaix.snapcast.utils.Setup.copyBinAsset(Setup.java:64)
03-20 16:41:49.651 32365  1721 E Setup   :      at de.badaix.snapcast.MainActivity$2.run(MainActivity.java:149)
03-20 16:41:49.651 32365  1721 E Setup   :      at java.lang.Thread.run(Thread.java:764)
03-20 16:41:49.961   872 11472 E NotificationService: Suppressing notification from package by user request.
03-20 16:41:50.966   872 16858 E NotificationService: Suppressing notification from package by user request.
03-20 16:41:51.637   872  1495 E NotificationService: Suppressing notification from package by user request.
03-20 16:41:51.672   872  1495 E NotificationService: Suppressing notification from package by user request.
03-20 16:41:51.675 32365 32365 E AndroidRuntime: FATAL EXCEPTION: main
03-20 16:41:51.675 32365 32365 E AndroidRuntime: Process: de.badaix.snapcast, PID: 32365

Thank you for your time and the great program.

**Comment by badaix** [comment]
I cannot really see why the App crashes. The two exceptions seem to be properly handled. 
Can you paste a complete logcat log for the de.phl.badaix process?
Might the supressed notification be the reason?
BTW: I've tested an X86 Oreo image in the emulator and it's running fine. Even the x86 snapclient binary is running in the emulator (althoug not completely in sync) :)

**Comment by badaix** [comment]
@gustavokch While updating my Android Studio, I found a deprecation warning about a notifications on Oreo. Maybe this was the reason for your crash: [Snapcast_0.13.1.apk.zip](https://github.com/badaix/snapcast/files/1865528/Snapcast_0.13.1.apk.zip)

**Comment by badaix** [comment]
fixed in release [v0.14.0](https://github.com/badaix/snapcast/releases/tag/v0.14.0)
